### PR TITLE
Implement support for Client Side Token Generation

### PR DESCRIPTION
**Pull request by IanDBird**
To test, refer to https://atlassian.thetradedesk.com/confluence/pages/viewpage.action?pageId=282321742

### CODE REVIEW

File: sdk/src/main/java/com/uid2/UID2Manager.kt [hunk 1]
```diff
@@ -155,6 +155,22 @@ public class UID2Manager internal constructor(
         }
     }
 
+    /**
+     * Generates a new identity with the given (hashed) email address.
+     *
+     * Once set, assuming it's valid, it will be monitored so that we automatically refresh the token(s) when required.
+     * This will also be persisted locally, so that when the application re-launches, we reload this Identity.
+     */
+    public fun generateIdentity(emailHash: String): Unit = afterInitialized {
```
Comment (sunnywu) [inline sdk/src/main/java/com/uid2/UID2Manager.kt:164]: just wondering, is it worth renaming to setIdentityFromEmail or similar to stay inline with the JS SDK convention?
https://unifiedid.com/docs/guides/publisher-client-side#configure-the-uid2-sdk-for-javascript

File: dev-app/src/main/java/com/uid2/dev/ui/MainScreenViewModel.kt [hunk 2]
```diff
@@ -77,18 +80,39 @@ class MainScreenViewModel(
...
+                    _viewState.emit(LoadingState)
+
                     try {
-                        // For Development purposes, we are required to generate the initial Identity before then
-                        // passing it onto the SDK to be managed.
-                        _viewState.emit(LoadingState)
-                        api.generateIdentity(action.address, EMAIL)?.let {
-                            manager.setIdentity(it)
+                        if (action.clientSide) {
+                            // Generate the identity via client side token generation.
```
Comment (sunnywu) [inline dev-app/src/main/java/com/uid2/dev/ui/MainScreenViewModel.kt:87]: _Generate the identity via Client Side Integration (client side token generation)_

?

File: dev-app/src/main/java/com/uid2/dev/ui/MainScreenViewModel.kt [hunk 3]
```diff
@@ -77,18 +80,39 @@ class MainScreenViewModel(
...
+                    } catch (ex: UID2Exception) {
                         _viewState.emit(ErrorState(ex))
                     }
                 }
+                is MainScreenAction.PhoneChanged -> {
+                    _viewState.emit(LoadingState)
 
+                    try {
+                        if (action.clientSide) {
+                            // Generate the identity via client side token generation.
```
Comment (sunnywu) [inline dev-app/src/main/java/com/uid2/dev/ui/MainScreenViewModel.kt:104]: same here?

File: sdk/src/main/java/com/uid2/UID2Client.kt [hunk 4]
```diff
@@ -20,27 +26,98 @@ import java.net.URL
...
     @Throws(
         InvalidApiUrlException::class,
-        RefreshTokenException::class,
+        CryptoException::class,
+        RequestFailureException::class,
+        PayloadDecryptException::class,
+        InvalidPayloadException::class,
+    )
+    suspend fun generateIdentity(identityRequest: IdentityRequest): UID2Identity? = withContext(ioDispatcher) {
+        // If the SDK hasn't been configured with the required client side token generation parameters, then it's not
```
Comment (sunnywu) [inline sdk/src/main/java/com/uid2/UID2Client.kt:59]: probably can keep it

File: sdk/src/main/java/com/uid2/UID2Manager.kt [hunk 5]
```diff
@@ -400,6 +439,12 @@ public class UID2Manager internal constructor(
         private const val UID2_API_URL_KEY = "uid2_api_url"
         private const val UID2_API_URL_DEFAULT = "https://prod.uidapi.com"
 
+        // The metadata keys that are used to provide access to required parameters for client side token generation.
```
Comment (sunnywu) [inline sdk/src/main/java/com/uid2/UID2Manager.kt:442]: _The metadata keys that are used to provide access to required parameters for Client Side Integration_ 
?

File: sdk/src/main/java/com/uid2/data/IdentityRequest.kt [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    val payloadValue = when (this) {
+        is Email -> data.toSha256()
+        is Phone -> data.toSha256()
+        else -> data
+    }
+
+    return JSONObject().apply {
+        put(payloadKey, payloadValue)
+
+        // Request that we check if the identity has opted out previously.
```
Comment (sunnywu) [inline sdk/src/main/java/com/uid2/data/IdentityRequest.kt:57]: Sorry maybe reword to:
_If the identity has opted out, we will get an optout response_

File: sdk/src/main/java/com/uid2/network/DataEnvelope.kt [hunk 7]
```diff
@@ -12,21 +13,19 @@ import javax.crypto.spec.SecretKeySpec
  * **See Also:**
  * [GitHub](https://github.com/IABTechLab/uid2docs/blob/main/api/v2/getting-started/gs-encryption-decryption.md)
  */
-public object DataEnvelope {
-    // The name and transformation of the encryption algorithm used.
-    private const val ALGORITHM_NAME = "AES"
-    private const val ALGORITHM_TRANSFORMATION = "AES/GCM/NoPadding"
+public class DataEnvelope {
```
Comment (mrmans0n) [inline sdk/src/main/java/com/uid2/network/DataEnvelope.kt:16]: Why did this become a class instead of an object? by looking at it, it seems it holds no state

File: sdk/src/main/java/com/uid2/utils/InputUtils.kt [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    private enum class EmailParsingState {
+        Starting,
+        SubDomain,
+    }
+
+    /**
+     * This code will attempt to normalize a given email address. It's been translated from a Java reference, and
+     * therefore been kept as close to the original implementation as possible (with some unused conditions removed).
+     *
+     * https://github.com/IABTechLab/uid2-operator/blob/main/src/main/java/com/uid2/operator/service/InputUtil.java#L96
```
Comment (mrmans0n) [inline sdk/src/main/java/com/uid2/utils/InputUtils.kt:86]: nit: This line reference is pointing to the main commit, but it is susceptible to change in the repo and it won't point to what you intend then. You might point to a specific commit so the line will always point to what you want.

File: sdk/src/main/java/com/uid2/utils/KeyUtils.kt [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+import java.security.spec.X509EncodedKeySpec
+import javax.crypto.KeyAgreement
+import javax.crypto.SecretKey
+import javax.crypto.spec.SecretKeySpec
+
+/**
+ * A class containing utility methods for generating Security based keys, e.g. [PublicKey], [KeyPair] and [SecretKey].
+ * This is useful to allow these methods to be easily mocked for unit tests.
+ */
+internal class KeyUtils {
```
Comment (mrmans0n) [inline sdk/src/main/java/com/uid2/utils/KeyUtils.kt:20]: should this be an object instead? or do you see value in being able to have multiple KeyUtils instances?

File: sdk/src/main/java/com/uid2/UID2Client.kt [hunk 10]
```diff
@@ -20,34 +26,105 @@ import java.net.URL
...
+    private val apiSubscriptionId: String?,
     private val session: NetworkSession,
+    private val packageName: String,
+    private val dataEnvelope: DataEnvelope = DataEnvelope(),
+    private val timeUtils: TimeUtils = TimeUtils(),
+    private val keyUtils: KeyUtils = KeyUtils(),
     private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
+    private val paramsFactory: (Map<String, String>) -> String = { params ->
+        JSONObject().apply {
+            params.forEach { (key, value) -> put(key, value) }
```
Comment (mrmans0n) [inline sdk/src/main/java/com/uid2/UID2Client.kt:39]: nit: up to you, you could use the JSONObject that takes a map
```suggestion
        JSONObject(params)
```

File: sdk/src/main/java/com/uid2/UID2Manager.kt [hunk 11]
```diff
@@ -161,6 +166,42 @@ public class UID2Manager internal constructor(
...
+     *
+     * Once set, assuming it's valid, it will be monitored so that we automatically refresh the token(s) when required.
+     * This will also be persisted locally, so that when the application re-launches, we reload this Identity.
+     *
+     * @throws InputValidationException Thrown if the given [IdentityRequest] is not valid. For a
+     * [IdentityRequest.Phone] we expect the given number to conform to the ITU E.164 Standard
+     * (https://en.wikipedia.org/wiki/E.164).
+     */
+    @Throws(InputValidationException::class)
+    public fun generateIdentity(identityRequest: IdentityRequest): Unit = afterInitialized {
```
Comment (koalaparadise) [inline sdk/src/main/java/com/uid2/UID2Manager.kt:180]: Does the caller need to know when the generation finished? At the moment this launches a coroutine so the caller will never know.

File: sdk/src/main/java/com/uid2/utils/TimeUtils.kt [hunk 12]
```diff
@@ -5,6 +5,11 @@ package com.uid2.utils
  */
 internal class TimeUtils {
 
+    /**
+     * Returns the current epoch time (in milliseconds).
+     */
+    fun now() = System.currentTimeMillis()
```
Comment (koalaparadise) [inline sdk/src/main/java/com/uid2/utils/TimeUtils.kt:11]: This should be `@JvmStatic`?

File: sdk/src/main/java/com/uid2/UID2Client.kt (hunk 10 above)
Comment (IanDBird) [reply sdk/src/main/java/com/uid2/UID2Client.kt:41]: Good point, thanks

File: sdk/src/main/java/com/uid2/utils/InputUtils.kt (hunk 8 above)
Comment (IanDBird) [reply sdk/src/main/java/com/uid2/utils/InputUtils.kt]: Sounds good, thanks

File: sdk/src/main/java/com/uid2/network/DataEnvelope.kt (hunk 7 above)
Comment (IanDBird) [reply sdk/src/main/java/com/uid2/network/DataEnvelope.kt:16]: `DataEnvelope` is mocked in `UID2ClientTest` so that I can return fake data when asked to decrypt.

File: sdk/src/main/java/com/uid2/utils/KeyUtils.kt (hunk 9 above)
Comment (IanDBird) [reply sdk/src/main/java/com/uid2/utils/KeyUtils.kt:20]: `KeyUtils` is a class so that I can mock, to avoid UTs calling crypto specific code.

File: sdk/src/main/java/com/uid2/utils/TimeUtils.kt (hunk 12 above)
Comment (IanDBird) [reply sdk/src/main/java/com/uid2/utils/TimeUtils.kt:11]: This is in a class (which is mocked) so can't be `@JvmStatic`, right? As this current version allows me to control what `now` is within the UT.

Comment (koalaparadise) [reply sdk/src/main/java/com/uid2/utils/TimeUtils.kt:11]: sgtm

File: sdk/src/main/java/com/uid2/UID2Manager.kt (hunk 11 above)
Comment (koalaparadise) [reply sdk/src/main/java/com/uid2/UID2Manager.kt:186]: Yeah I'd add an `onResult: (Result) -> Unit` parameter. `Result` being a simple sealed class with Error, Success, etc

File: sdk/src/main/java/com/uid2/network/DataEnvelope.kt (hunk 7 above)
Comment (IanDBird) [reply sdk/src/main/java/com/uid2/network/DataEnvelope.kt]: Addressed with commits shared by Nacho!

File: sdk/src/main/java/com/uid2/utils/KeyUtils.kt (hunk 9 above)
Comment (IanDBird) [reply sdk/src/main/java/com/uid2/utils/KeyUtils.kt]: Addressed with commits shared by Nacho!

**Review by mrmans0n — APPROVED** [review]

File: sdk/src/main/java/com/uid2/UID2Client.kt [hunk 13]
```diff
@@ -22,34 +27,116 @@ import java.net.URL
...
+            NetworkRequestType.POST,
+            mapOf(),
+            paramsFactory(
+                mapOf(
+                    "payload" to encryptedPayload.encodeBase64(),
+                    "iv" to iv.encodeBase64(),
+                    "public_key" to clientKeyPair.public.encoded.encodeBase64(),
+                    "timestamp" to now.toString(),
+                    "subscription_id" to subscriptionId,
+                    "app_name" to packageName,
```
[the remark this answers is not in the document]
Comment (IanDBird) [reply sdk/src/main/java/com/uid2/UID2Client.kt:104]: The package name should **always** already be lower case if provided correctly. That said, there is no harm in formatting it into lower case when we obtain it. I'll look to do this separately (at the source).

File: dev-app/src/main/AndroidManifest.xml [hunk 14]
```diff
@@ -13,7 +13,8 @@
         android:supportsRtl="true"
         tools:targetApi="31">
 
-        <!-- The metadata consumed by the Dev App + SDK -->
+        <!-- Metadata Required for Server Side Integration -->
```
[the remark this answers is not in the document]
Comment (IanDBird) [reply dev-app/src/main/AndroidManifest.xml:16]: Thanks for the suggestion, will fix!
